// xavren/dist/src/index.js
"use strict";
// dotenv-cloud/index.js
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// import fs from 'fs';
// import path from 'path';
// import WebSocket from 'ws';
// import crypto from 'crypto';
// const fs = require("fs");
// const chokidar = require("chokidar");
const fs_1 = __importDefault(require("fs"));
const chokidar_1 = __importDefault(require("chokidar"));
// const watcher = chokidar.watch(".env", { persistent: true });
// watcher.on("change", (path) => {
//   console.log(`${path} changed, syncing with cloud DB...`);
//   // 👉 trigger your sync
// });
function watchEnv({ filePath = ".env", onChange }) {
    const watcher = chokidar_1.default.watch(filePath, { persistent: true });
    watcher.on("change", (path) => {
        console.log(`${path} changed, syncing with cloud DB...`);
        if (onChange) {
            const content = fs_1.default.readFileSync(path, "utf8"); // read file content
            onChange(content, path);
        }
    });
    return watcher; // return watcher in case you want to close it later
}
watchEnv({ onChange: (content) => {
        console.log(content);
    } });
// function encrypt(text, key, secret) {
//   const cipher = crypto.createCipher('aes-256-cbc', key + secret);
//   let encrypted = cipher.update(text, 'utf8', 'hex');
//   encrypted += cipher.final('hex');
//   return encrypted;
// }
// function decrypt(text, key, secret) {
//   const decipher = crypto.createDecipher('aes-256-cbc', key + secret);
//   let decrypted = decipher.update(text, 'hex', 'utf8');
//   decrypted += decipher.final('utf8');
//   return decrypted;
// }
// function parseEnvFile(envPath) {
//   if (!fs.existsSync(envPath)) return {};
//   const raw = fs.readFileSync(envPath, 'utf8');
//   const lines = raw.split('\n');
//   const parsed = {};
//   for (const line of lines) {
//     if (!line || line.startsWith('#')) continue;
//     const [k, ...v] = line.split('=');
//     parsed[k.trim()] = v.join('=').trim();
//   }
//   return parsed;
// }
// function writeProcessEnv(parsedEnv, key, secret) {
//   Object.keys(parsedEnv).forEach(k => {
//     try {
//       process.env[k] = decrypt(parsedEnv[k], key, secret);
//     } catch {
//       process.env[k] = parsedEnv[k];
//     }
//   });
// }
// export function config({ env, onChange, key, secret }) {
//   const envPath = env || path.resolve(process.cwd(), '.env');
//   let envVars = parseEnvFile(envPath);
//   writeProcessEnv(envVars, key, secret);
//   // WebSocket connection for real-time sync
//   const ws = new WebSocket('wss://your-dotenv-cloud-server.com');
//   ws.on('open', () => {
//     console.log('Connected to dotenv-cloud server');
//     const encryptedEnv = encrypt(JSON.stringify(envVars), key, secret);
//     ws.send(JSON.stringify({ type: 'init', env: encryptedEnv }));
//   });
//   ws.on('message', message => {
//     try {
//       const data = JSON.parse(message);
//       if (data.type === 'update') {
//         const updatedEnv = JSON.parse(decrypt(data.env, key, secret));
//         const changedEnv = {};
//         Object.keys(updatedEnv).forEach(k => {
//           if (envVars[k] !== updatedEnv[k]) {
//             changedEnv[k] = updatedEnv[k];
//           }
//         });
//         envVars = { ...envVars, ...updatedEnv };
//         writeProcessEnv(envVars, key, secret);
//         if (onChange) onChange({ env: envVars, changedEnv });
//       }
//     } catch (err) {
//       console.error('Failed to process update from server:', err);
//     }
//   });
//   // Watch local env file changes and sync to server
//   fs.watch(envPath, (eventType) => {
//     if (eventType === 'change') {
//       const newEnv = parseEnvFile(envPath);
//       const changedEnv = {};
//       Object.keys(newEnv).forEach(k => {
//         if (envVars[k] !== newEnv[k]) changedEnv[k] = newEnv[k];
//       });
//       if (Object.keys(changedEnv).length === 0) return;
//       envVars = { ...envVars, ...newEnv };
//       writeProcessEnv(envVars, key, secret);
//       if (onChange) onChange({ env: envVars, changedEnv });
//       // Send updated env to server
//       const encryptedEnv = encrypt(JSON.stringify(envVars), key, secret);
//       ws.send(JSON.stringify({ type: 'update', env: encryptedEnv }));
//     }
//   });
// }
