// xavren/dist/cryption.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleDecrypt = handleDecrypt;
exports.deriveKey = deriveKey;
exports.handleEncryptEnv = handleEncryptEnv;
exports.handleDecryptEnv = handleDecryptEnv;
const crypto_1 = __importDefault(require("crypto"));
const ALGORITHM = "aes-256-gcm";
function handleDecrypt({ encrypted: encryptedData, passphrase, }) {
    let { salt: saltHex, iv: ivHex, authTag: tagHex, encrypted } = encryptedData;
    encrypted = encrypted || encryptedData?.key;
    const salt = Buffer.from(saltHex, "hex");
    const iv = Buffer.from(ivHex, "hex");
    const tag = Buffer.from(tagHex, "hex");
    const key = crypto_1.default.scryptSync(passphrase, salt, 32);
    const decipher = crypto_1.default.createDecipheriv(ALGORITHM, key, iv);
    decipher.setAuthTag(tag);
    let decrypted = decipher.update(encrypted, "hex", "utf8");
    decrypted += decipher.final("utf8");
    // return tagHex
    return decrypted;
}
function deriveKey(passphrase, salt) {
    return crypto_1.default.scryptSync(passphrase, salt, 32);
}
function handleEncryptEnv({ data: obj, passphrase: secretPhrase, }) {
    // Generate a random salt
    const salt = crypto_1.default.randomBytes(16);
    // Derive a key from secret phrase + salt
    const key = deriveKey(secretPhrase, salt);
    // Extract key-value
    const title = obj.title;
    const value = obj.value;
    // Generate random IV for each field
    const iv1 = crypto_1.default.randomBytes(16);
    const iv2 = crypto_1.default.randomBytes(16);
    // Encrypt title
    const cipher1 = crypto_1.default.createCipheriv("aes-256-gcm", key, iv1);
    const encryptedTitle = Buffer.concat([
        cipher1.update(title, "utf8"),
        cipher1.final(),
    ]);
    const authTag1 = cipher1.getAuthTag();
    // Encrypt value
    const cipher2 = crypto_1.default.createCipheriv("aes-256-gcm", key, iv2);
    const encryptedValue = Buffer.concat([
        cipher2.update(value, "utf8"),
        cipher2.final(),
    ]);
    const authTag2 = cipher2.getAuthTag();
    return {
        title: encryptedTitle.toString("hex"),
        value: encryptedValue.toString("hex"),
        salt: salt.toString("hex"),
        ivTitle: iv1.toString("hex"),
        ivValue: iv2.toString("hex"),
        authTagTitle: authTag1.toString("hex"),
        authTagValue: authTag2.toString("hex"),
        // _id:obj?.item?._id
    };
}
function handleDecryptEnv({ title, value, salt, ivTitle, ivValue, authTagTitle, authTagValue, }, passphrase) {
    const saltBuf = Buffer.from(salt, "hex");
    const key = deriveKey(passphrase, saltBuf);
    // --- Decrypt title ---
    const decipherTitle = crypto_1.default.createDecipheriv(ALGORITHM, key, Buffer.from(ivTitle, "hex"));
    decipherTitle.setAuthTag(Buffer.from(authTagTitle, "hex"));
    let decryptedTitle = decipherTitle.update(Buffer.from(title, "hex"));
    decryptedTitle = Buffer.concat([decryptedTitle, decipherTitle.final()]);
    // --- Decrypt value ---
    const decipherValue = crypto_1.default.createDecipheriv(ALGORITHM, key, Buffer.from(ivValue, "hex"));
    decipherValue.setAuthTag(Buffer.from(authTagValue, "hex"));
    let decryptedValue = decipherValue.update(Buffer.from(value, "hex"));
    decryptedValue = Buffer.concat([decryptedValue, decipherValue.final()]);
    return {
        title: decryptedTitle.toString(),
        value: decryptedValue.toString(),
        //     item:{
        //     title,
        //   value,
        //   salt,
        //   ivTitle,
        //   ivValue,
        //   authTagTitle,
        //   authTagValue,
        // }
    };
}
