// xavren/dist/state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.envStore = exports.sessionStore = void 0;
const vanilla_1 = require("zustand/vanilla");
// Create store without React
exports.sessionStore = (0, vanilla_1.createStore)((set) => ({
    user: null,
    project: null,
    token: null,
    email: null,
    secretPhrase: null,
    config: null,
    login: (user, token) => set({ user, token }),
    logout: () => set({ user: null, token: null }),
    setSecretPhrase: (secretPhrase) => set({ secretPhrase }),
    setEmail: (email) => set({ email }),
    setProject: (project) => set({ project }),
    setConfig: (config) => set({ config }),
}));
// Vanilla zustand store
exports.envStore = (0, vanilla_1.createStore)((set, get) => ({
    envs: [],
    addEnv: (env) => set((state) => ({
        envs: [...state.envs, env],
    })),
    setEnvs: (env) => set({
        envs: env,
    }),
    removeEnv: (index) => set((state) => ({
        envs: state.envs.filter((_, i) => i !== index),
    })),
    clearEnvs: () => set({ envs: [] }),
    getEnv: (index) => {
        const envs = get().envs;
        return envs[index] || null;
    },
}));
